notify/factory.go

```diff
@@ -27,57 +31,147 @@ import (
 	"github.com/grafana/alerting/receivers/wecom"
 )
 
-var receiverFactories = map[string]func(receivers.FactoryConfig) (NotificationChannel, error){
-	"prometheus-alertmanager": wrap(alertmanager.New),
-	"dingding":                wrap(dinding.New),
-	"discord":                 wrap(discord.New),
-	"email":                   wrap(email.New),
-	"googlechat":              wrap(googlechat.New),
-	"kafka":                   wrap(kafka.New),
-	"line":                    wrap(line.New),
-	"opsgenie":                wrap(opsgenie.New),
-	"pagerduty":               wrap(pagerduty.New),
-	"pushover":                wrap(pushover.New),
-	"sensugo":                 wrap(sensugo.New),
-	"slack":                   wrap(slack.New),
-	"teams":                   wrap(teams.New),
-	"telegram":                wrap(telegram.New),
-	"threema":                 wrap(threema.New),
-	"victorops":               wrap(victorops.New),
-	"webhook":                 wrap(webhook.New),
-	"wecom":                   wrap(wecom.New),
-	"webex":                   wrap(webex.New),
-}
-
-type NotificationChannel interface {
-	notify.Notifier
-	notify.ResolvedSender
-}
-
-func Factory(receiverType string) (func(receivers.FactoryConfig) (NotificationChannel, error), bool) {
-	receiverType = strings.ToLower(receiverType)
-	factory, exists := receiverFactories[receiverType]
-	return factory, exists
-}
-
-// wrap wraps the notifier's factory errors with receivers.ReceiverValidationError
-func wrap[T NotificationChannel](f func(fc receivers.FactoryConfig) (T, error)) func(receivers.FactoryConfig) (NotificationChannel, error) {
-	return func(fc receivers.FactoryConfig) (NotificationChannel, error) {
-		ch, err := f(fc)
-		if err != nil {
-			return nil, ReceiverValidationError{
-				Err: err,
-				// TODO it will be removed in the next PR
-				Cfg: &GrafanaReceiver{
-					UID:                   fc.Config.UID,
-					Name:                  fc.Config.Name,
-					Type:                  fc.Config.Type,
-					DisableResolveMessage: fc.Config.DisableResolveMessage,
-					Settings:              fc.Config.Settings,
-					SecureSettings:        nil,
-				},
+// BuildReceiverIntegrations creates integrations for each configured notification channel in GrafanaReceiverConfig.
+// It returns a slice of Integration objects, one for each notification channel, along with any errors that occurred.
+func BuildReceiverIntegrations(
+	receiver GrafanaReceiverConfig,
+	tmpl *template.Template,
```

Comment:
```suggestion
	tmpl *alerting.Template,
```

you need the local types otherwise you'll import the alertmanage directly.

Reply:
I will address it in a follow-up PR as it touches many lines.